server/aesdsocket.c

```diff
@@ -0,0 +1,319 @@
+	sa.sa_mask = blocked_signals;
+	sa.sa_flags = 0;
+	sigaction(SIGTERM, &sa, NULL);
+	sigaction(SIGINT, &sa, NULL);
+
+	//References: AESD course slides "Sockets pg. 15: Getting Sockaddr", "https://beej.us/guide/bgnet: pg. 21"
+	//The below was pulled almost directly from above resources
+ 	
+ 	//Get address info for localhost port 9000
+ 	struct addrinfo hints;
+ 	memset(&hints, 0, sizeof(hints));
+ 	
+ 	hints.ai_flags    = AI_PASSIVE;
+ 	hints.ai_socktype = SOCK_STREAM;
+ 	hints.ai_family   = AF_INET6;
+ 	
+ 	struct addrinfo *ai_result;
+ 	
+ 	int rc = getaddrinfo(NULL, "9000", &hints, &ai_result);
+ 	
+ 	if (rc != 0){
+ 		syslog((LOG_USER | LOG_INFO),"Error obtaining address info!");
+ 		return -1;
+ 	}
+
+
+ 	//bind address to socket
+ 	//reference:https://beej.us/guide/bgnet/pdf/bgnet_usl_c_1.pdf pg. 35
+ 	int socket_fd;
+ 	struct addrinfo *current_node; //current node of LL containing addr info results
+ 	
+ 	//Traverse linked list looking for valid address info to create and bind socket
+ 	for (current_node = ai_result; current_node != NULL; current_node=current_node->ai_next){
+ 		socket_fd = socket(AF_INET6, SOCK_STREAM, 0);
+		if (socket_fd == -1){
+			syslog((LOG_USER | LOG_INFO),"socket creation failed, trying next");
+			continue;
+		}
+		
+		int reuse_value=1; //boolean value to set SO_REUSEADDR: 
+		rc = setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &reuse_value, sizeof(reuse_value));
+		if (rc == -1){
+			syslog((LOG_USER | LOG_INFO),"Error setting socket address to reusable!");
+			return -1;
+		}
+		
+		rc = bind(socket_fd, current_node->ai_addr, current_node->ai_addrlen);
+		if (rc == -1){
+			close(socket_fd);
+ 			syslog((LOG_USER | LOG_INFO),"socket bind failed, trying next");
+ 			continue;
+ 		}
+ 		break;
+ 	}
+ 	
+ 	freeaddrinfo(ai_result); //no longer need address info after binding
+ 	
+ 	if (current_node == NULL){
+ 		syslog((LOG_USER | LOG_INFO),"Error binding socket! No socket binded, exiting program...");
+		return -1;
+ 	}
+ 	else{
+ 		syslog((LOG_USER | LOG_INFO),"Socket successfully binded!");
+ 		if (argc > 1){
+ 			if (!strcmp(argv[1],"-d")){
+ 				//we are goin demon mode
+				if ( daemon(0,1) == -1){
+					syslog((LOG_USER | LOG_INFO),"Error binding socket! No socket binded, exiting program...");
+					return -1;
+				}
+				
+ 			}
+ 		}
+ 	}
+ 	
+ 	//listen for connections
+ 	rc = listen(socket_fd, 5); //wait for connection, allow backlog of 5 waiters...
+ 	if (rc == -1){
+ 		syslog((LOG_USER | LOG_INFO),"Error listening to socket");
+ 		return -1;
+ 	}
+ 		
+ 	//Open or Create file for input data
+ 	int packetdata_fd = open("/var/tmp/aesdsocketdata", (O_RDWR | O_CREAT | O_APPEND), 0644);
+ 	if (packetdata_fd == -1){
+ 		syslog((LOG_USER | LOG_INFO),"Errror creating/opening temp data file");
+ 		return -1;
+ 	}
+	
+	//TODO: This does introduce a race condition around signal flag if it is updated between times where it is checked.
+	//It would be better to poll on a signal fd. Can try if I have time.
```

Comment:
You can also use `shutdown` on the `socket_fd` - this should allow a blocking accept to exit.  See the slide from class about this.